Change casting of utsname.Machine to fix make binary for ARM

StefanScherer opened the pull request:
The PR #17478 breaks `make binary` for ARM.

On a Scaleway machine we now get this error:

```
pkg/platform/architecture_linux.go:15: cannot use utsname.Machine (type [65]uint8) as type [65]int8 in argument to charsToString
```

The reason is that utsname is defined differently for ARM and AMD64:
- AMD64: https://golang.org/src/syscall/ztypes_linux_amd64.go?s=9792:9941#L560 
- ARM: https://golang.org/src/syscall/ztypes_linux_arm.go?#L542

To make it work again we need to cast it correctly for both architectures.

This PR is also needed to continue with the work on **Support Docker on ARM** #17802

StefanScherer committed:
Move charsToString to architecture dependent source to fix casting problem

Signed-off-by: Stefan Scherer <scherer_stefan@icloud.com>

calavera commented:
We're going to make the distinction by architecture, now it fails in i386:

```
pkg/platform/architecture_linux.go:15: cannot convert utsname.Machine (type [65]int8) to type [65]uint8
```

StefanScherer commented:
@calavera I'll fix my PR.

StefanScherer commented:
@calavera I wanted to avoid duplicating code with this cast. Now I just splitted out the `charsToString()` function into architecture specific go sources.

I squashed and pushed my changes twice, so I don't know if the Jenkins build actually work on the latest push.

calavera commented:
Tests are running properly, those failures are flakey. Thanks.

LGTM

calavera commented:
@StefanScherer can we add godoc for those functions saying why we have both?

StefanScherer commented:
@calavera Good point. I'll add some docs later.

StefanScherer commented:
Just clarified that we have to live with such differences on different architectures.
https://github.com/golang/go/issues/13318

I'll improve the PR to use the +build tags so we only need two files instead of many more. Otherwise I have to add more copies of the architecture_linux_amd64.go source.

StefanScherer commented:
@calavera I've updated the PR, added some comments where to find the different types of syscall Utsname in the GO sources. I renamed the sources and now use +build tags to select one of the two implementations for three architectures as I found this usage in GO 1.5.1 at the moment:

```
$ grep -r Machine .
./src/syscall/ztypes_linux_386.go:       Machine    [65]int8
./src/syscall/ztypes_linux_amd64.go:     Machine    [65]int8
./src/syscall/ztypes_linux_arm.go:       Machine    [65]uint8
./src/syscall/ztypes_linux_arm64.go:     Machine    [65]int8
./src/syscall/ztypes_linux_ppc64.go:     Machine    [65]uint8
./src/syscall/ztypes_linux_ppc64le.go:   Machine    [65]uint8
```

So this should probably help @brahmaroutu for ppc64le as well.

calavera commented:
oh gosh, this sounds like so much fun.

LGTM

vdemeester commented:
oh waw :sweat_smile: 
LGTM 🐻

StefanScherer commented:
Yes, it was fun ;-)

brahmaroutu commented:
@StefanScherer thanks for letting me know. This looks good for ppc64le. Please add 's390x' for utsname_uint8.go since we support IBM Z. (or I can do that after your PR is merged)
I tested with the following line on Z and it compiles and unit tests run fine. 
// +build linux,arm linux,ppc64 linux,ppc64le s390x

StefanScherer commented:
@brahmaroutu I've added it.

brahmaroutu commented:
@StefanScherer thank you.

StefanScherer commented:
@vdemeester Hope this last commit was fine for you. But it just adds the comment for @brahmaroutu additional architecture.

vdemeester commented:
@StefanScherer yeah no problem, still LGTM :+1:

StefanScherer commented:
Thanks! We'll send you more ARM related PR's later on.
